Account wide ability via config file and improvements.

BoiseComputer opened the pull request:
**NOTE:** This was originally made to update the account wide version from: https://github.com/SoulSeekkor/mod-reagent-bank-account
He was unresponsive to responding to my pull request, so I am offering my updated version here. Let me know if you want me to implement my version directly into your base version here.

---

This pull request introduces significant updates and improvements to the `mod-reagent-bank-account` module, including enhanced functionality, better configuration options, and improved documentation. The changes focus on expanding features, standardizing SQL paths and table names, and refining the codebase for maintainability and compatibility.

### Documentation and Feature Enhancements:
* Expanded the `README.md` with detailed installation, configuration, and usage instructions, as well as a changelog and credits section. Added new features such as reagent withdrawal in stack sizes, NPC banker interaction, and support for all trade goods and gems.

### SQL Path and Table Standardization:
* Updated SQL file paths in `conf.sh.dist` to use a standardized directory structure (`data/sql/db-auth`, `data/sql/db-characters`, `data/sql/db-world`).
* Renamed and updated SQL files to use consistent table and script names (`mod_reagent_bank_account`), replacing the old `custom_reagent_bank_account` naming convention. [[1]](diffhunk://#diff-b7d9239f1d7938004803e82ad06b152bc7b0444342506389c7613c8a78153b4bR1-R7) [[2]](diffhunk://#diff-cc32c1d0384926729648dbd8e67f62e9443245934738c265a800ea98393213f5L1-L8) [[3]](diffhunk://#diff-65363ac1d63f1a24ce2c2d8809ac12287c3d6a66136611e3b09e8f88088c6d5cR1-R9) [[4]](diffhunk://#diff-65fc190d24ce0464c30a2e474ff00c96c1219cca9c753465ec88cb4802975c5dL1-L9)

### Configuration Improvements:
* Renamed `reagent_bank_account.conf.dist` to `mod_reagent_bank_account.conf.dist` and added a new configuration option (`ReagentBankAccount.MaxOptionsPerPage`) to control the number of items displayed per page in the NPC menu.

### Codebase Updates:
* Introduced a global variable `g_maxOptionsPerPage` in `ReagentBankAccount.h` to dynamically manage the maximum number of options per page, replacing the hardcoded `MAX_OPTIONS`.
* Simplified the loader script in `ReagentBankAccount_loader.cpp` by removing unnecessary comments.

SoulSeekkor committed:
Added config setting for account-wide.

SoulSeekkor committed:
Rearranged menus and updated to order returned items by descending order.

SoulSeekkor committed:
Moved the 'Back' button to the top.

SoulSeekkor committed:
Added query to add account_id column and update it.

SoulSeekkor committed:
Cleaned up config.

SoulSeekkor committed:
Revert "Added config setting for account-wide."

This reverts commit 6a3b806168ea0d71304737287321c943ec66274a.

SoulSeekkor committed:
Updated to make the module a completely separate module.

SoulSeekkor committed:
Updated to use account_id instead of character_id.

SoulSeekkor committed:
Updated method call.

SoulSeekkor committed:
Renamed files to make them unique.

SoulSeekkor committed:
Updated to use a new class name.

SoulSeekkor committed:
Updated another spot for the class name.

SoulSeekkor committed:
Updated to also have next/previous at the bottom of the page.

SoulSeekkor committed:
Updated to output items deposited.

SoulSeekkor committed:
Updated to perform proper conversions.

SoulSeekkor committed:
Updated chat message calls.

SoulSeekkor committed:
Updated to include another back button at the bottom and to move the chat message.

SoulSeekkor committed:
Updated to fix chat message.

SoulSeekkor committed:
Another chat message fix.

SoulSeekkor committed:
Removed NPC tag.

SoulSeekkor committed:
Updated to clean up menus and to now have to scroll.

SoulSeekkor committed:
Remove chat message for counts.

SoulSeekkor committed:
Rearranged some menu options.

SoulSeekkor committed:
Updated to include chat messages for what was deposited.

SoulSeekkor committed:
Fixed not using existing count and wrong map variable name.

SoulSeekkor committed:
Moved code block to proper place.

SoulSeekkor committed:
Updated to make the deposited messages more concise.

SoulSeekkor committed:
Updated so next page is always on the top.

SoulSeekkor committed:
Moved 'Other Trade Goods' menu up higher.

BoiseComputer committed:
Update conf.sh.dist

BoiseComputer committed:
Update reagent_bank_account_NPC.sql

BoiseComputer committed:
Update reagent_bank_account_NPC.sql

BoiseComputer committed:
Rename reagent_bank_account.conf.dist to mod_reagent_bank_account.conf.dist

BoiseComputer committed:
Refactor reagent bank module: update SQL table structure, improve NPC script, and enhance README documentation

BoiseComputer committed:
Refactor reagent bank module: improve configuration file, enhance NPC script, and add new withdrawal functionality

BoiseComputer committed:
Update NPC SQL script: replace REPLACE statements with DELETE and INSERT for better clarity and functionality

BoiseComputer committed:
Added ability to switch from per character to account wide.

Badgermilk0 commented:
Do you have your own branch availible for download until someone merges? :)

BoiseComputer commented:
> Do you have your own branch availible for download until someone merges? :)
Feel free to test out this version: https://github.com/BoiseComputer/mod-reagent-bank-account/tree/master

AldebaraanMKII commented:
@BoiseComputer
Script named 'mod_reagent_bank_account' is not assigned in the database.
this is what i get when i clone your fork. I added all sqls file to their respective directories.

BoiseComputer commented:
> @BoiseComputer Script named 'mod_reagent_bank_account' is not assigned in the database. this is what i get when i clone your fork. I added all sqls file to their respective directories.

Make sure that you are applying this is the characters DB.
Should be this: https://github.com/BoiseComputer/mod-reagent-bank-account/blob/master/data/sql/db-characters/base/mod_reagent_bank_account_create_table.sql

AldebaraanMKII commented:
> > @BoiseComputer Script named 'mod_reagent_bank_account' is not assigned in the database. this is what i get when i clone your fork. I added all sqls file to their respective directories.
> 
> Make sure that you are applying this is the characters DB. Should be this: https://github.com/BoiseComputer/mod-reagent-bank-account/blob/master/data/sql/db-characters/base/mod_reagent_bank_account_create_table.sql

```
/* Note: (1050) Table 'mod_reagent_bank_account' already exists */
/* Affected rows: 0  Found rows: 0  Warnings: 1  Duration for 1 query: 0,000 sec. */
```
problem is that it already exists. For some unknown reason the error still occurs.

I remade the database from scratch by placing all sql files from the azerothcore repo into the data folder, and the issue still occurs.

mod_reagent_bank_account_NPC.sql is in "data\sql\custom\db_world\base"
mod_reagent_bank_account_create_table.sql is in "data\sql\custom\db_characters\base"

When i talk to the bank NPC i get no menu to deposit the items.

BoiseComputer commented:
> > > @BoiseComputer Script named 'mod_reagent_bank_account' is not assigned in the database. this is what i get when i clone your fork. I added all sqls file to their respective directories.
> > 
> > 
> > Make sure that you are applying this is the characters DB. Should be this: https://github.com/BoiseComputer/mod-reagent-bank-account/blob/master/data/sql/db-characters/base/mod_reagent_bank_account_create_table.sql
> 
> ```
> /* Note: (1050) Table 'mod_reagent_bank_account' already exists */
> /* Affected rows: 0  Found rows: 0  Warnings: 1  Duration for 1 query: 0,000 sec. */
> ```
> 
> problem is that it already exists. For some unknown reason the error still occurs.
> 
> I remade the database from scratch by placing all sql files from the azerothcore repo into the data folder, and the issue still occurs.
> 
> mod_reagent_bank_account_NPC.sql is in "data\sql\custom\db_world\base" mod_reagent_bank_account_create_table.sql is in "data\sql\custom\db_characters\base"
> 
> When i talk to the bank NPC i get no menu to deposit the items.

Here is where mine is: ![image](https://github.com/user-attachments/assets/18762d1a-7b15-411a-a03b-8ef297120b13)

LuckyLuc96 commented:
Hi @BoiseComputer I installed your module, but I was not able to see it in action. I had no errors during the installation and I imported both your SQL scripts into the proper databases (acore_characters) and (acore_world). 

But the NPC does not show any of the custom options when spawned in.

I also made sure to drop the old tables from the earlier versions of this module. 

![image](https://github.com/user-attachments/assets/293a9f6b-e048-4a86-bdb9-9aba5f73a1c2)

edit:

I actually get this error message when running worldserver

"Script named 'npc_reagent_banker' is assigned in the database, but has no code!"

edit2:

I found the problem I just gotta figure out how to link these two up. It's a naming issue on my side.

"Script named 'mod_reagent_bank_account' is not assigned in the database.
Script named 'npc_reagent_banker' is assigned in the database, but has no code!"

Fixed it with this query:
UPDATE creature_template SET ScriptName = 'mod_reagent_bank_account' WHERE entry = 290011;

BoiseComputer commented:
I'll fix the script and get this patched. It is most likely a setting I did locally, but it didn't make it into the SQL template.

BoiseComputer committed:
Implement account-wide reagent bank functionality with enhanced withdrawal options

- Introduced account-wide storage keys for managing reagents based on account or character.
- Added methods for withdrawing items: single item, stack, and all items.
- Enhanced item withdrawal menu to include options for withdrawing one item, a full stack, or all items.
- Improved item deposit functionality to handle multiple categories and provide feedback to players.
- Refactored code for better readability and maintainability, including consistent use of lambdas for database queries.
- Added caching for item templates and icons to optimize performance.
- Implemented pagination for displaying stored reagents, allowing players to navigate through multiple pages of items.

BoiseComputer committed:
Refactor SQL query syntax in reagent bank module for improved readability

BoiseComputer committed:
Refactor message formatting in reagent bank withdrawal functions for improved clarity and consistency

ZhengPeiRu21 commented:
Hello! I apologize for taking so long to respond; I am on hiatus from AC development these days. It looks like there are some merge conflicts, but if you wouldn't mind updating this PR to resolve them, I'd be happy to merge it. Thanks for your contribution!

PurpleTonberry commented:
@BoiseComputer I wasn't sure where else to put this. I'm using the fork of reagent bank account with your tweaks, I assume it's more or less the same as your changes here? The "Deposit All" option inside the individual categories is unsafe. If the item already exists in the reagent banker, it is overwritten by the value of what you just put in.

Seems to happen in every category and is simple to reproduce.

Have stuff in there already, go to a category, select an item you have a lot of, take 1. While still inside that same category, select Deposit All. Re-open the banker, and the item count should now be 1.
